### PR TITLE
Feature/issue 1465 standalone

### PR DESCRIPTION
**Pull request by wds15**
## Summary

This introduces a small makefile which will help users of the Stan Math C++ library to compile programs on their own. The `make/standalone` makefile makes building of the math libraries straightforward and allows building of C++ easily using the make facility.

The `README.md` is updated to use the new standalone makefile in order to demonstrate building a small hello world program which got a little more complicated since the 3.0 release as the TBB must be linked in dynamically which is now being handled automatically through make.

## Tests

NA

## Side Effects

NA

## Checklist

- [X] Math issue #1465

- [X] Copyright holder: Sebastian Weber

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ Documentation for Stan math is available at [mc-stan.org/math](https://mc-stan.o
 
 Installation
 ------------
-The Stan Math Library is a C++ library which depends on the Intel TBB library and requires for some functionality (ordinary differential equations and root solving) on the Sundials library.
+The Stan Math Library is a C++ library which depends on the Intel TBB library and requires for some functionality (ordinary differential equations and root solving) on the Sundials library. As build system the make facility is used to manage all dependencies.
```
Comment (mcol) [inline README.md]: I would say: "As build system, we use the make facility to manage all dependencies."

File: README.md [hunk 2]
```diff
@@ -45,12 +45,13 @@ If this is in the file `/path/to/foo/foo.cpp`, then you can compile and run this
...
-> clang++ -std=c++1y -I /path/to/stan-math -I /path/to/Eigen -I /path/to/boost -I /path/to/sundials -I /path/to/tbb -L /path/to/tbb-libs -ltbb -D_REENTRANT  foo.cpp
-> ./a.out
+> make -f path/to/stan-math/make/standalone math-libs
+> make -f path/to/stan-math/make/standalone foo
+> ./foo
 log normal(1 | 2, 3)=-2.07311
 ```
 
-The `-I` includes provide paths pointing to the five necessary includes:
+The first make command with the `math-libs` target ensures that all binary dependencies are build and ready to use. The second make command ensures that all dependencies of Stan Math are available to the compiler which are:
```
Comment (mcol) [inline README.md]: "... all binary dependencies are built ..."

Split the second sentence in two: "The second make command ensures that all dependencies of Stan Math are available to the compiler. These are:"

File: README.md [hunk 3]
```diff
@@ -60,11 +61,16 @@ The `-I` includes provide paths pointing to the five necessary includes:
...
 
+```bash
+> make -f ~/stan-dev/math/make/standalone math-libs
+> make -f ~/stan-dev/math/make/standalone foo
 ```
-clang++ -std=c++1y -I ~/stan-dev/math -I ~/stan-dev/math/lib/eigen_3.3.3/ -I ~/stan-dev/math/lib/boost_1.69.0/ -I ~/stan-dev/math/lib/sundials_4.1.0/include  -I ~/stan-dev/math/lib/tbb_2019_U8/include -L ~/stan-dev/math/lib/tbb -ltbb -Wl,-rpath,"~/stan-dev/math/lib/tbb" -D_REENTRANT foo.cpp
-```
 
-The following directories all exist below the links given to `-I`: `~/stan-dev/math/stan` and `~/stan-dev/math/lib/eigen_3.3.3/Eigen` and `~/stan-dev/math/lib/boost_1.69.0/boost` and `~/stan-dev/math/lib/sundials_4.1.0/include` and `~/stan-dev/math/lib/tbb_2019_U8/include`. The `~/stan-dev/math/lib/tbb` directory is created by the stan-math makefiles automatically when running any unit test (for example with `./runTests.py test/unit/math/rev/core/agrad_test.cpp`). The `-Wl,-rpath,...` instruct the linker to hard-code the path to the Intel TBB library inside the stan-math directory into the final binary. This way the Intel TBB is found when executing the program.
+The `math-libs` target must only be called a single time and can be omitted for subsequent compilations.
```
Comment (mcol) [inline README.md]: "The `math-libs` target has to be called only once, and ..."

File: README.md [hunk 4]
```diff
@@ -60,11 +61,16 @@ The `-I` includes provide paths pointing to the five necessary includes:
...
+
+The standalone makefile ensure that all the required `-I` include statements are given to the compiler and the necessary libraries are linked: `~/stan-dev/math/stan` and `~/stan-dev/math/lib/eigen_3.3.3/Eigen` and `~/stan-dev/math/lib/boost_1.69.0/boost` and `~/stan-dev/math/lib/sundials_4.1.0/include` and `~/stan-dev/math/lib/tbb_2019_U8/include`. The `~/stan-dev/math/lib/tbb` directory is created by the math-libs makefile target automatically. The `-Wl,-rpath,...` instruct the linker to hard-code the path to the Intel TBB library inside the stan-math directory into the final binary. This way the Intel TBB is found when executing the program.
```
Comment (mcol) [inline README.md]: "The standalone makefile ensures..."

"... by the `math-libs` makefile target ..."

File: README.md [hunk 5]
```diff
@@ -60,11 +61,16 @@ The `-I` includes provide paths pointing to the five necessary includes:
...
+
+Note for Windows users: On Windows the rpath feature as used by Stan Math to hardcode an absolute path to a dynamically loaded library does not work. On Windows the TBB dynamic library `tbb.dll` is located in the `math/lib/tbb` directory. The user can choose to either copy this file to the same directory of the executable or to add the directory `path/to/math/lib/tbb` as absolute path to the system wide `PATH` variable.
```
Comment (mcol) [inline README.md]: "... can choose either to copy ..."

"system-wide"

File: README.md (hunk 1 above)
Comment (wds15) [reply README.md]: I think we never use "we" in this file, so I would not start to use it, but have a look at the rewording.

**Comment by wds15** [comment]
Thanks for the suggestions. These should be addressed now... and I added `make/local` configuration support.

Good to go for me.

File: README.md (hunk 1 above)
Comment (mcol) [reply README.md]: I hadn't noticed that, but I agree with you.

File: README.md [hunk 6]
```diff
@@ -51,7 +51,7 @@ If this is in the file `/path/to/foo/foo.cpp`, then you can compile and run this
 log normal(1 | 2, 3)=-2.07311
 ```
 
-The first make command with the `math-libs` target ensures that all binary dependencies are build and ready to use. The second make command ensures that all dependencies of Stan Math are available to the compiler which are:
+The first make command with the `math-libs` target ensures that all binary dependencies are build and ready to use. The second make command ensures that all dependencies of Stan Math are available to the compiler. These are:
```
Comment (mcol) [inline README.md]: You missed to correct "are built".

File: README.md [hunk 7]
```diff
@@ -66,12 +66,20 @@ Note that the paths should *not* include the final directories `stan`, `Eigen`,
...
+Note for Windows users: On Windows the `-rpath` feature as used by Stan Math to hardcode an absolute path to a dynamically loaded library does not work. On Windows the Intel TBB dynamic library `tbb.dll` is located in the `math/lib/tbb` directory. The user can choose to copy this file to the same directory of the executable or to add the directory `path/to/math/lib/tbb` as absolute path to the system-wide `PATH` variable.
+
+Compilers
+---------
+The above example will use the default compiler of the system as determined by `make`. On Linux this is usually `g++`, on MacOS `clang++`, and for Windows this is `g++` if the RTools for Windows are used. There's nothing special about any of these and they can be changed through the `CXX` variable of `make`. The recommended way to set this variable for the Stan Math library is by creating a `path/to/math/make/local` file. Defining `CXX=g++` in this file will ensure that always the GNU C++ compiler is used, for example. The compiler must be able to fully support C++11 and partially the C++14 standard. The `g++` 4.9.3 version part of RTools for Windows currently defines the minimal C++ feature set required by the Stan Math library.
```
Comment (mcol) [inline README.md]: I think `path/to/math/make/local` could be confusing. This seems clearer:
"... is to create a `make/local` file within the Stan Math directory."

"... that the GNU C++ compiler is always used"

**Pull request merged**